src/systems/environmental_sensor_system/EnvironmentalSensorSystem.cc

```diff
@@ -76,19 +78,89 @@ class EnvironmentalSensor : public gz::sensors::Sensor
     // Load common sensor params
     gz::sensors::Sensor::Load(_sdf);
 
-    this->pub = this->node.Advertise<gz::msgs::Double>(this->Topic());
+    // Handle the field type
+    if (_sdf.Element() != nullptr &&
+      _sdf.Element()->HasElement("output_format"))
+    {
+      std::string format = _sdf.Element()->Get<std::string>("output_format");
+      if (format == "scalar")
+      {
+        this->numberOfFields = 1;
+      }
+      else if (format == "vector3")
+      {
+        this->numberOfFields = 3;
+      }
+      else
+      {
+        gzerr << "Invalid output format " << format << "given."
+          << "valid options are scalar or vector3. "
+          << "Defaulting to 1." << std::endl;
+      }
+    }
+
+    // Hansle the transform type
+    if (_sdf.Element() != nullptr &&
+      _sdf.Element()->HasElement("transform_type"))
+    {
+      auto res =
+        getTransformType(_sdf.Element()->Get<std::string>("transform_type"));
+
+      if (!res.has_value())
+      {
+        gzerr << "Invalid transform type of "
+          << _sdf.Element()->Get<std::string>("transform_type")
+          << " given. Valid types are ADD_VELOCITY_LOCAL, ADD_VELOCITY_GLOBAL,"
+          << " LOCAL, and GLOBAL. Defaulting to GLOBAL." << std::endl;
+      }
+      else
+      {
+        this->transformType = res.value();
+      }
+    }
+
+    switch(this->numberOfFields) {
+      case 1:
+        this->pub = this->node.Advertise<gz::msgs::Double>(this->Topic());
+        break;
+      case 3:
+        this->pub = this->node.Advertise<gz::msgs::Vector3d>(this->Topic());
+        break;
+      default:
+        gzerr << "Unable to create publisher too many fields" << "\n";
+    }
 
     // If "environment_variable" is defined then remap
     // sensor from existing data.
     if (_sdf.Element() != nullptr &&
+      this->numberOfFields == 1 &&
       _sdf.Element()->HasElement("environment_variable"))
```

Comment:
@arjo129 hmm, should this be `x_variable` instead? If we want to keep `environment_variable` for scalar data, I'd suggest `environment_variable_x` and the like for vector data.